### PR TITLE
feat: overhaul and streamline Android page navigation transitions

### PR DESCRIPTION
**Pull request by ADjenkov**
[Tracking issue](https://github.com/NativeScript/NativeScript/issues/7772)

### CODE REVIEW

**Comment by SvetoslavTsenov** [comment]
test --ignore testsapptransitions api23 ios12

**Comment by SvetoslavTsenov** [comment]
test --ignore tests_app_transitions api23 ios12

**Comment by ADjenkov** [comment]
test --ignore tests_app_transitions api23

**Comment by ADjenkov** [comment]
test --ignore ngapps_android animations_android api23 api28

**Comment by ADjenkov** [comment]
test --ignore ngapps_android#djenkov/comment-tests api23 api28

File: tns-platform-declarations/android/org.nativescript.widgets.d.ts [hunk 1]
```diff
@@ -173,12 +178,6 @@
                 public verticalAlignment: VerticalAlignment;
             }
 
-            export class FragmentBase extends androidx.fragment.app.Fragment {
```
Comment (manoldonev) [inline tns-platform-declarations/android/org.nativescript.widgets.d.ts]: Not sure removing this is a good idea:
1) Technically removing public api is a breaking change
2) We actually duplicate the functionality of `getRemovingParentFragment()` on many places in the new version

File: tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+    public CustomTransition(AnimatorSet animatorSet, String transitionName) {
+        this.animatorSet = animatorSet;
+        this.transitionName = transitionName;
+    }
+
+    @Nullable
+    @Override
+    public Animator onAppear(@NonNull ViewGroup sceneRoot, @NonNull final View view, @Nullable TransitionValues startValues,
+                             @Nullable TransitionValues endValues) {
+        if (endValues == null || view == null || this.animatorSet == null) {
```
Comment (manoldonev) [inline tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java]: What scenario is this? Can we add a comment?

File: tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+    public String getTransitionName(){
+            return this.transitionName;
+    }
+
+    private Animator setAnimatorsTarget(AnimatorSet animatorSet, final View view){
+        ArrayList<Animator> animatorsList = animatorSet.getChildAnimations();
+        boolean resetOnTransitionEnd = this.resetOnTransitionEnd;
+        
+        for (int i = 0; i < animatorsList.size(); i++) {
+                animatorsList.get(i).setTarget(view);
```
Comment (manoldonev) [inline tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java]: Formatting.

File: tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        if (this.resetOnTransitionEnd) {
+                this.immediateAnimatorSet = this.animatorSet.clone();
```
Comment (manoldonev) [inline tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java]: Formatting.

File: tns-core-modules/ui/bottom-navigation/bottom-navigation.android.ts [hunk 5]
```diff
@@ -77,18 +79,61 @@ function initializeNativeClasses() {
         public onCreate(savedInstanceState: android.os.Bundle): void {
             super.onCreate(savedInstanceState);
             const args = this.getArguments();
-            this.tab = getTabById(args.getInt(TABID));
+            this.owner = getTabById(args.getInt(TABID));
             this.index = args.getInt(INDEX);
-            if (!this.tab) {
-                throw new Error(`Cannot find BottomNavigation`);
+            if (!this.owner) {
+                throw new Error(`Cannot find TabView`);
```
Comment (manoldonev) [inline tns-core-modules/ui/bottom-navigation/bottom-navigation.android.ts]: Cannot find BottomNavigation

File: tns-core-modules/ui/bottom-navigation/bottom-navigation.android.ts [hunk 6]
```diff
@@ -77,18 +79,61 @@ function initializeNativeClasses() {
...
+            const tabItem = this.owner.items[this.index];
 
             return tabItem.nativeViewProtected;
         }
+
+        public onDestroyView() {
+            const hasRemovingParent = this.getParentFragment() && this.getParentFragment().isRemoving();
+
+            // Get view as bitmap and set it as background. This is workaround for the disapearing nested fragments.
+            // TO DO: Consider removing it when update to androidx.fragment:1.2.0
```
Comment (manoldonev) [inline tns-core-modules/ui/bottom-navigation/bottom-navigation.android.ts]: // TODO:

File: tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        CustomAnimatorListener listener = new CustomAnimatorListener(view);
+        animatorSet.addListener(listener);
+        addListener(new CustomTransitionListenerAdapter(this));
```
Comment (manoldonev) [inline tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java]: this.addListener

File: tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+        
+        for (int i = 0; i < animatorsList.size(); i++) {
+                animatorsList.get(i).setTarget(view);
+        }
+
+        if (this.resetOnTransitionEnd) {
+                this.immediateAnimatorSet = this.animatorSet.clone();
+        }
+
+        CustomAnimatorListener listener = new CustomAnimatorListener(view);
```
Comment (manoldonev) [inline tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java]: Add comment for hardware acceleration.

File: tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    private class CustomTransitionListenerAdapter extends TransitionListenerAdapter {
+        private CustomTransition customTransition;
+
+        CustomTransitionListenerAdapter(CustomTransition transition) {
+            this.customTransition = transition;
+        }
+
+        @Override
+        public void onTransitionEnd(@NonNull Transition transition) {
+            if (this.customTransition.resetOnTransitionEnd) {
```
Comment (manoldonev) [inline tns-core-modules-widgets/android/widgets/src/main/java/org/nativescript/widgets/CustomTransition.java]: Add comment

File: tns-core-modules/ui/bottom-navigation/bottom-navigation.android.ts [hunk 10]
```diff
@@ -77,18 +79,61 @@ function initializeNativeClasses() {
...
+            // const bitmap = android.graphics.Bitmap.createBitmap(width, height, android.graphics.Bitmap.Config.ARGB_8888);
+            // const canvas = new android.graphics.Canvas(bitmap);
+            // view.layout(0, 0, width, height);
+            // view.draw(canvas);
+
+            view.setDrawingCacheEnabled(true);
+            const bitmap = android.graphics.Bitmap.createBitmap(view.getDrawingCache());
+            view.setDrawingCacheEnabled(false);
+
+            return bitmap;
```
Comment (manoldonev) [inline tns-core-modules/ui/bottom-navigation/bottom-navigation.android.ts]: Can this return null? Would `new android.graphics.drawable.BitmapDrawable(appResources, null)` throw an exception?

File: tns-core-modules/ui/bottom-navigation/bottom-navigation.android.ts [hunk 11]
```diff
@@ -77,18 +79,61 @@ function initializeNativeClasses() {
...
+            // TO DO: Consider removing it when update to androidx.fragment:1.2.0
+            if (hasRemovingParent && this.owner.selectedIndex === this.index) {
+                this.backgroundBitmap = this.loadBitmapFromView(this.owner.nativeViewProtected);
+            }
+
+            super.onPause();
+        }
+
+        private loadBitmapFromView(view: android.view.View): android.graphics.Bitmap {
+            // Another way to get view bitmap. Test performance vs setDrawingCacheEnabled
```
Comment (manoldonev) [inline tns-core-modules/ui/bottom-navigation/bottom-navigation.android.ts]: TODO:

File: tns-core-modules/ui/frame/fragment.transitions.android.ts [hunk 12]
```diff
@@ -121,114 +92,172 @@ export function _setAndroidFragmentTransitions(
...
             currentFragmentNeedsDifferentAnimation = true;
         }
     }
 
     if (name === "none") {
-        transition = new NoTransition(0, null);
+        const noTransition = new NoTransition(0, null);
+
+        if (isNestedDefaultTransition) {
+            fragmentTransaction.setCustomAnimations(android.R.anim.fade_in, android.R.anim.fade_out);
```
Comment (manoldonev) [inline tns-core-modules/ui/frame/fragment.transitions.android.ts]: Add comment for this special case.

File: tns-core-modules/ui/tab-view/tab-view.android.ts [hunk 13]
```diff
@@ -70,18 +72,61 @@ function initializeNativeClasses() {
...
+            // TO DO: Consider removing it when update to androidx.fragment:1.2.0
+            if (hasRemovingParent && this.owner.selectedIndex === this.index) {
+                this.backgroundBitmap = this.loadBitmapFromView(this.owner.nativeViewProtected);
+            }
+
+            super.onPause();
+        }
+
+        private loadBitmapFromView(view: android.view.View): android.graphics.Bitmap {
+            // Another way to get view bitmap. Test performance vs setDrawingCacheEnabled
```
Comment (manoldonev) [inline tns-core-modules/ui/tab-view/tab-view.android.ts]: Unify with the other implementation

File: tns-platform-declarations/android/org.nativescript.widgets.d.ts (hunk 1 above)
Comment (ADjenkov) [reply tns-platform-declarations/android/org.nativescript.widgets.d.ts]: I agree now

**Comment by ADjenkov** [comment]
test --ignore tsapps_android tscapps ngapps_android api23 api28

**Pull request merged**